### PR TITLE
Setup port-forwarding for linkerd dashboard command

### LINKED ISSUE #1696: use port-forward for linkerd dashboard
The `linkerd dashboard` command uses a kubernetes proxy to allow the user to connect from their laptop to the Linkerd dashboard and Grafana.

Instead, port-forward from the user's machine. This has several advantages:
- simple tcp forwarding vs. http proxy
- kubernetes proxy inspects and rewrites http
- Linkerd dashboard and Grafana may be served at `localhost:1234/` root, rather than a long proxy URL, this would help solve the `root_url` Grafana issue discussed in #1692

### PR DESCRIPTION
**Pull request by klingerf**
This branch switches the `linkerd dashboard` command from using the kubernetes service proxy to access the web UI, to instead using port-forward to access the web pod directly. This simplifies the URLs displayed in the dashboard, and it removes our direct dependency on "k8s.io/kubernetes" 🎉 

As part of this change, I fixed a bug with `linkerd dashboard`, where the proxy would not start until after the dashboard had been successfully loaded in a browser (relates to #2047). And I've updated the integration tests to use port-forwarding instead of proxying to access pods running in the test environment.

Fixes #1696.

### CODE REVIEW

**Comment by grampelberg** [comment]
Woah, now that's an awesome PR.

**Comment by klingerf** [comment]
Note that there's still a lot of code in the `web/` directory for loading the web UI through the service proxy. If we decide that we're not going to support viewing the web UI via `kubectl proxy`, then we could rip all of that out and it would clean things up a lot. @grampelberg, do you have a preference one way or another?

**Comment by grampelberg** [comment]
@klingerf I'm totally behind removing support for viewing the dashboard it via `kubectl proxy`.

File: cli/cmd/dashboard.go [hunk 1]
```diff
@@ -20,19 +21,25 @@ const (
 
 	// showURL displays dashboard URLs without opening a browser.
 	showURL = "url"
+
+	// webDeployment is the name of the web UI's deployment in the install config
+	webDeployment = "linkerd-web"
+
+	// webPort is the port on which the web UI is served in the web pod
+	webPort = 8084
```
Comment (siggy) [inline cli/cmd/dashboard.go]: mind adding a note that `webDeployment` and `webPort` must match values in `template.go`? perhaps eventually these should be shared constants.

File: pkg/k8s/portforward.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/portforward"
```
Comment (siggy) [inline pkg/k8s/portforward.go]: 👍

File: pkg/k8s/portforward.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	}
+
+	req := clientset.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Namespace(namespace).
+		Name(podName).
+		SubResource("portforward")
+
+	if localPort == 0 {
+		localPort, err = getLocalPort()
```
Comment (siggy) [inline pkg/k8s/portforward.go]: is there a potential timing issue where we call `getLocalPort` in `NewPortForward`, and then that `localPort` is taken before `Run()` is called?

is the reason we're calling `getLocalPort` here to enable `URLFor` to work immediately after calling `NewPortForward`?

File: pkg/k8s/portforward.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+func (pf *PortForward) Run() error {
+	transport, upgrader, err := spdy.RoundTripperFor(pf.config)
+	if err != nil {
+		return err
+	}
+
+	ports := []string{fmt.Sprintf("%d:%d", pf.localPort, pf.remotePort)}
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, pf.method, pf.url)
+
+	fw, err := portforward.New(dialer, ports, pf.stopCh, pf.readyCh, ioutil.Discard, ioutil.Discard)
```
Comment (siggy) [inline pkg/k8s/portforward.go]: consider `os.Stdout, os.Stderr`, perhaps when `verbose` is enabled?

File: Gopkg.toml [hunk 5]
```diff
@@ -60,26 +56,18 @@ required = [
...
-
 [[override]]
   name = "github.com/json-iterator/go"
   revision = "f2b4162afba35581b6d4a50d3b8f34e33c144682"
 
 [[override]]
   name = "github.com/Azure/go-autorest"
   revision = "1ff28809256a84bb6966640ff3d0371af82ccba4"
+
+[[override]]
```
[the remark this answers is not in the document]
Comment (klingerf) [reply Gopkg.toml]: This is a result of kubernetes/apimachinery#46 / golang/dep#1207. We could try upgrading our k8s.io deps to latest and removing these overrides, but I didn't want to open up that can of worms.

File: pkg/k8s/portforward.go [hunk 6]
```diff
@@ -103,10 +107,17 @@ func (pf *PortForward) Run() error {
...
+	if pf.emitLogs {
+		out = os.Stdout
+		errOut = os.Stderr
+	}
+
 	ports := []string{fmt.Sprintf("%d:%d", pf.localPort, pf.remotePort)}
 	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, pf.method, pf.url)
 
-	fw, err := portforward.New(dialer, ports, pf.stopCh, pf.readyCh, ioutil.Discard, ioutil.Discard)
+	fw, err := portforward.New(dialer, ports, pf.stopCh, pf.readyCh, out, errOut)
```
Comment (siggy) [inline pkg/k8s/portforward.go]: nice!

```bash
Forwarding from 127.0.0.1:51328 -> 8084
Forwarding from [::1]:51328 -> 8084
Linkerd dashboard available at:
http://127.0.0.1:51328
Grafana dashboard available at:
http://127.0.0.1:51328/grafana
Handling connection for 51328
```

**Pull request merged**